[JENKINS-55681] - Make JAXB a detached plugin only on Java 11

batmat opened the pull request:
[JENKINS-55681](https://issues.jenkins-ci.org/browse/JENKINS-55681)

* First commit refactors ClassicPluginStrategy to move the detached plugins logic into a dedicated DetachedPluginsManager class
* Second commit introduces code to allow a plugin to be detached only if the Java runtime greather or equal to a given version.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* [JENKINS-55681](https://issues.jenkins-ci.org/browse/JENKINS-55681): Support running Jenkins WARs on Java 11 without downloading JAXB JARs. A new JAXB plugin is marked as a detached one when running with Java 11 or higher.
  * https://plugins.jenkins.io/jaxb

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@jenkinsci/java11-support 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

batmat committed:
Move detached plugins logic into a dedicated class

ClassicPluginStrategy is already long enough, so trying to make things a
bit more focused before adding Java 11 logic. (and even if not, the change
still splits a class that is almost 1k lines).

batmat committed:
Install jaxb as detached only on Java 11+

jglick commented:
Try fiddling with `core/src/main/resources/jenkins/split-plugin-cycles.txt`? Or implement JENKINS-28942 already?

jglick commented on core/src/main/resources/jenkins/split-plugin-cycles.txt:
```diff
@@ -10,3 +10,7 @@ script-security command-launcher
 script-security windows-slaves
+script-security jaxb
+command-launcher jaxb
+script-security jaxb
```
Duplicate?

batmat replied:
Yeah, :man_facepalming:

batmat commented:
Given the PR was conflicted, I chose to rebase and force-push it while I was marking it a non-WIP anymore. This is ready for review.

oleg-nenashev commented on core/src/main/java/jenkins/plugins/DetachedPluginsManager.java:
```diff
@@ -0,0 +1,227 @@
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class DetachedPluginsManager {
```
The name looks confusing since it does not implement "PluginManager". Add "Helper" to the name?

batmat replied:
I see the point. It's more an old habit of calling things that "manage" something a Thing__Manager__. Or Handler. _Helper_ I'm _slightly_ uncomfortable with, but will think about it.

oleg-nenashev commented on core/src/main/resources/jenkins/split-plugin-cycles.txt:
```diff
@@ -10,3 +10,6 @@ script-security command-launcher
 script-security matrix-auth
 script-security matrix-project
 script-security windows-slaves
+script-security jaxb
```
Would it apply to any detached plugin?

batmat replied:
I'm quite unclear about why this seemed to surface only for JAXB TBH. Through some local testing, it seems like it can be random (?!?) because I wasn't able to reproduce the `StackOverflowError` locally, when this was visible in CI (before I added these lines to `split-plugin-cycles.txt`)

oleg-nenashev replied:
It seems to be YOLO in the current state. I suggest adding break cycles for all previously detached plugins

oleg-nenashev replied:
Still needs discussion

batmat replied:
Done @oleg-nenashev

oleg-nenashev reviewed (commented):
Need to test it manually before approving. Why did you need a massive refactoring to the utility class in this PR?

MRamonLeon commented on core/src/main/java/jenkins/plugins/DetachedPluginsManager.java:
```diff
@@ -93,34 +101,32 @@ private DetachedPluginsManager() {
     public static @Nonnull
     List<DetachedPlugin> getDetachedPlugins() {
-        return DETACHED_LIST;
+        return DETACHED_LIST.stream()
```
Can't this be done in the initialization of the DETACHED_LIST directly?  If possible, we don't need extra changes along the class changing DETACHED_LIST by getDetachedPlugins

batmat replied:
Yeah, I considered it already, not 100% sure why I chose this path. I'll have another look.

MRamonLeon commented on core/src/main/java/jenkins/plugins/DetachedPluginsManager.java:
```diff
@@ -0,0 +1,214 @@
+     *
+     * @return A {@link List} of {@link DetachedPlugin}s.
+     */
+    @Restricted(NoExternalUse.class)
```
is it needed as all the class is restricted? (the same with some other methods)

batmat replied:
`Restricted` should generally be the default behavior when adding code. And yes, that is intended because I don't wish to make it a public API forever :-).

MRamonLeon replied:
I meant, the whole class is already Restricted. Do you need to annotate the method too? Isn't it enough with the annotation on the class?

batmat replied:
done, removed from methods. I left it also on the internal class, just in case the access-modifier handles this incorrectly. Should be harmful in any case.

MRamonLeon reviewed (commented):
A couple of comments

batmat committed:
Remove duplicate @Restricted

Given the enclosing class already has it.

jglick commented on core/src/test/java/hudson/ClassicPluginStrategyTest.java:
```diff
@@ -38,20 +39,20 @@
     @Test
     public void test_getDetachedPlugins() {
-        List<DetachedPlugin> list = ClassicPluginStrategy.getDetachedPlugins(new VersionNumber("1.296"));
+        List<DetachedPlugin> list = DetachedPluginsManager.getDetachedPlugins(new VersionNumber("1.296"));
```
Then perhaps just move this to a test suite named after the class it is testing. Or at a minimum use `@For`.

batmat committed:
Use 1.5 release

batmat committed:
Merge remote-tracking branch 'origin/master' into detach-jaxb-on-Java-11

oleg-nenashev reviewed (changes requested):
https://github.com/jenkinsci/jenkins/pull/3865/files#r251290617 needs to be resolved somehow. I also do not think that a massive refactoring was needed in this PR, but it is too late. Generally it's better to have it in a separate PR

oleg-nenashev commented on core/src/main/java/jenkins/plugins/DetachedPluginsManager.java:
```diff
@@ -0,0 +1,224 @@
+        }
+
+        @Nonnull
+        public VersionNumber getMinJavaVersion() {
```
```suggestion
        public VersionNumber getMinimumJavaVersion() {
```

batmat committed:
Add more test coverage

batmat committed:
getMinJavaVersion => getMinimumJavaVersion

batmat committed:
Rename DetachedPluginsManager as DetachedPluginsUtil

as requested by Oleg and Jesse. Mainly because of the misleading
relationship (there's no inheritance) with the existing `PluginManager`
class.

batmat committed:
:facepalm: adjust to 1.5 release method renaming

batmat commented:
So for the record, in addition to the existing tests, my manual testing does confirm this work, at least on a fresh instance. I just confirmed the following works perfectly:
* Start a fresh Jenkins instance using a build of Jenkins WAR for this PR, on Java 11 runtime
* install SlocCount plugin as usual from the plugin manager UI, see that JAXB gets installed automatically, despite the plugin does not declare JAXB dependency
* Use the plugin, and see that there's not ClassNotFoundException anymore

What I plan to confirm works too, and seems critical, but didn't test yet:

### upgrade on previous version of runtime & Jenkins

* run on Java 8 a previous Jenkins version, configure the job and check this works fine
* update to Java 11 on the WAR for this PR
* check the JAXB plugin gets installed and the job still works

### upgrade of Java runtime only
* Run PR's WAR on Java 8, configure the job, etc.
* Upgrade Java runtime to 11 only
* check the JAXB plugin gets installed and the job still works

batmat commented:
For anyone interested, I recorded a video demonstrating the effect of this PR: https://www.youtube.com/watch?v=OWfurCu-ODU

See also https://github.com/batmat/jaxb-java11-demo#use-this-image for an easy way to test it yourself

oleg-nenashev commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -864,6 +705,5 @@ protected Class defineClassFromData(File container, byte[] classData, String cla
     }
 
     public static boolean useAntClassLoader = SystemProperties.getBoolean(ClassicPluginStrategy.class.getName()+".useAntClassLoader");
-    private static final Logger LOGGER = Logger.getLogger(ClassicPluginStrategy.class.getName());
```
CC @varyvol @fcojfernandez  This change likely invalidates the logging tweak in https://github.com/jenkinsci/jenkinsfile-runner/pull/49 / JENKINS-54425 . I am not sure which logger is used in line 692 after this removal

oleg-nenashev commented on core/src/main/resources/jenkins/split-plugins.txt:
```diff
@@ -1,6 +1,6 @@
 # See ClassicPluginStrategy.DETACHED_LIST. As of JENKINS-47634 also used by plugin-compat-tester.
 
-# Columns are: plugin ID, last core release still containing the plugin's functionality, plugin version that's implied.
+# Columns are: plugin ID, last core release still containing the plugin's functionality, plugin version that's implied, min jdk version where to install the plugin
```
```suggestion
# Columns are: plugin ID, last core release still containing the plugin's functionality, plugin version that's implied, minimum Java specification version where to install the plugin
```

oleg-nenashev commented on core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:
```diff
@@ -0,0 +1,224 @@
+                return new DetachedPluginsUtil.DetachedPlugin(pieces[0],
+                                                              pieces[1] + ".*",
+                                                              pieces[2],
+                                                              pieces.length == 4 ? pieces[3] : "1.0");
```
```suggestion
                                                              pieces.length >= 4 ? pieces[3] : "1.0");
```

Just in case there is a trailing data in the resource file

batmat replied:
I don't really see how this could happen. I mean, I would 100% agree for an external config file, but this file is internal and is very rarely changed. 
I guess if I was to do a change here, that would be more adding a test to check there's max 4 columns, and fail if not.

batmat replied:
Done as https://github.com/jenkinsci/jenkins/pull/3865/commits/a0e38ee9f8e93000177c609c8a5ea11050f0933e

oleg-nenashev reviewed (commented):
Looks solid && almost ready to go, thanks @batmat . I added few comments/proposals

oleg-nenashev committed:
Clarify minimum Java version column

Co-Authored-By: batmat <bmathus@gmail.com>

batmat committed:
Update to 2.3.0 version of the jaxb plugin

batmat committed:
Bump minimum core version to next expected release version

batmat committed:
Merge branch 'master' into detach-jaxb-on-Java-11

Also to trigger the creation of an Incremental release.

batmat committed:
Add preventive cycle breaks for jaxb while waiting for JENKINS-28942

As requested by Oleg.

oleg-nenashev reviewed (approved):
Looks good to me. I will create a follow to JFR to restore the logging supression after this change is landed

alecharp reviewed (approved):
just one / two questions of code style but overall LGTM.

batmat commented on core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:
```diff
@@ -0,0 +1,224 @@
+     * @see JavaUtils#getCurrentJavaRuntimeVersionNumber()
+     */
+    public static @Nonnull
+    List<DetachedPlugin> getDetachedPlugins() {
```
Cut and paste then reformatted automatically indeed.

jglick replied:
Sounds like there was a bug in someone’s IDE, may as well correct it now.

MRamonLeon reviewed (approved):
Good job!

fcojfernandez reviewed (approved):
LGTM

batmat committed:
Add test to fail fast on wrong format for split-plugins.txt

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -266,32 +257,16 @@ private static void fix(Attributes atts, List<PluginWrapper.Dependency> optional
-     * @since 2.0
+     * @see DetachedPluginsUtil#getImpliedDependencies(String, String)
      */
+    @Deprecated // since TODO
```
We usually use `FIXME` for planned `@since` tags (easier to search for), but anyway I think there is no need to give a date/version on this. Just replace `@see` with a `@deprecated` tag.

jglick commented on core/src/main/java/hudson/ClassicPluginStrategy.java:
```diff
@@ -483,7 +324,7 @@ public void initializeComponents(PluginWrapper plugin) {
          */
         if (LOGGER.isLoggable(Level.FINER))
-            LOGGER.log(Level.FINER,"Scout-loading ExtensionList: "+type, new Throwable());
+            LOGGER.log(Level.FINER, "Scout-loading ExtensionList: "+type, new Throwable());
```
?

batmat replied:
:man_facepalming: unrelated ide reformatting I'll have missed I think ..

jglick commented on core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:
```diff
@@ -0,0 +1,224 @@
+package jenkins.plugins;
```
license header

jglick commented on core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:
```diff
@@ -0,0 +1,224 @@
+ * <p>
+ * This code was originally moved from {@link ClassicPluginStrategy}.
+ *
+ * @since TODO
```
Omit for a `@Restricted` class—it is nobody’s business when it happened to be added.

jglick commented on core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    /**
+     * Get the list of all plugins that have ever been {@link DetachedPlugin detached} from Jenkins core, applicable to the current Java runtime.
```
`@linkplain`?

jglick commented on core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:
```diff
@@ -0,0 +1,224 @@
+    /**
+     * Get the list of all plugins that have ever been {@link DetachedPlugin detached} from Jenkins core, applicable to the current Java runtime.
+     *
+     * @return A {@link List} of {@link DetachedPlugin}s.
```
Useless to just repeat the type signature of the method. If you have nothing new to add, delete the tag.

jglick commented on core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:
```diff
@@ -0,0 +1,224 @@
+     * </li>
+     * </ul>
+     */
+    @Restricted(NoExternalUse.class)
```
Unnecessary: the enclosing class was already restricted.

jglick commented on core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:
```diff
@@ -0,0 +1,224 @@
+         * Gets the minimum required version for the current version of Jenkins.
+         *
+         * @return the minimum required version for the current version of Jenkins.
+         * @since 2.16
```
delete

alecharp commented on core/src/test/java/jenkins/plugins/DetachedPluginsUtilTest.java:
```diff
@@ -0,0 +1,70 @@
+        assertFalse( "weird, split-plugins.txt only has comments?", linesWithoutComments.isEmpty());
+
+        //
+        assertFalse("no whitespaces only lines allowed" ,linesWithoutComments.stream()
```
```suggestion
        assertFalse("no whitespaces only lines allowed", linesWithoutComments.stream()
```

jglick commented on core/src/main/resources/jenkins/split-plugins.txt:
```diff
@@ -1,6 +1,6 @@
 # See ClassicPluginStrategy.DETACHED_LIST. As of JENKINS-47634 also used by plugin-compat-tester.
 
-# Columns are: plugin ID, last core release still containing the plugin's functionality, plugin version that's implied.
+# Columns are: plugin ID, last core release still containing the plugin's functionality, plugin version that's implied, minimum Java specification version where to install the plugin
```
Getting harder to read—would suggest moving to JSON but, no comments! And we have no YAML lib in `jenkins-core` AFAIK. Sigh…

jglick commented on core/src/test/java/hudson/ClassicPluginStrategyTest.java:
```diff
@@ -38,20 +39,20 @@
 
     @Test
     public void test_getDetachedPlugins() {
```
Does this just belong in `DetachedPluginsUtilTest` now?

alecharp commented on core/src/test/java/jenkins/plugins/DetachedPluginsUtilTest.java:
```diff
@@ -0,0 +1,70 @@
+                            .anyMatch(line -> !line.isEmpty()));
+
+
+        assertTrue( "max 4 columns is supported", linesWithoutComments.stream()
```
```suggestion
        assertTrue("max 4 columns is supported", linesWithoutComments.stream()
```

jglick commented on core/src/test/java/jenkins/plugins/DetachedPluginsUtilTest.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * Checks the format of the <code>/jenkins/split-plugins.txt</code> file has maximum 4 columns.
```
Could just enforce these things with `assert`ions while parsing.

jglick reviewed (approved):
Minor comments but looks good overall.

oleg-nenashev commented:
I have created https://issues.jenkins-ci.org/browse/JENKINS-55930 as a follow-up so that @jglick's comments can be addressed after the merge

Since there is no negative votes in the PR, I will proceed with merge. It will produce experimental Docker images which can be used for for testing of outstanding use-cases like upgradeability
